File: compilers/go/frontend/ast.go
// Package frontend implements passes 1-4 of the TSOP compiler in Go:
// parse, validate, typecheck, and ANF lowering.
package frontend

// ---------------------------------------------------------------------------
// Source locations
// ---------------------------------------------------------------------------

// SourceLocation represents a position in a source file.
type SourceLocation struct {
	File   string
	Line   int
	Column int
}

// ---------------------------------------------------------------------------
// Type nodes
// ---------------------------------------------------------------------------

// TypeNode represents a type annotation in the TSOP AST.
type TypeNode interface {
	typeNodeMarker()
}

// PrimitiveType is a built-in scalar type like bigint, boolean, ByteString, etc.
type PrimitiveType struct {
	Name string // e.g. "bigint", "boolean", "ByteString", "PubKey", ...
}

func (PrimitiveType) typeNodeMarker() {}

// FixedArrayType is a fixed-length array type: FixedArray<T, N>.
type FixedArrayType struct {
	Element TypeNode
	Length  int
}

func (FixedArrayType) typeNodeMarker() {}

// CustomType is an unrecognized type reference.
type CustomType struct {
	Name string
}

func (CustomType) typeNodeMarker() {}

// ---------------------------------------------------------------------------
// Top-level nodes
// ---------------------------------------------------------------------------

// ContractNode is the parsed representation of a TSOP smart contract class.
type ContractNode struct {
	Name        string
	ParentClass string // "SmartContract" or "StatefulSmartContract"
	Properties  []PropertyNode
	Constructor MethodNode
	Methods     []MethodNode
	SourceFile  string
}

// PropertyNode represents a contract property declaration.
type PropertyNode struct {
	Name           string
	Type           TypeNode
	Readonly       bool
	SourceLocation SourceLocation
}

// MethodNode represents a contract method or constructor.
type MethodNode struct {
	Name           string
	Params         []ParamNode
	Body           []Statement
	Visibility     string // "public" or "private"
	SourceLocation SourceLocation
}

// ParamNode represents a method parameter.
type ParamNode struct {
	Name string
	Type TypeNode
}

// ---------------------------------------------------------------------------
// Statements
// ---------------------------------------------------------------------------

// Statement is the interface for all AST statement nodes.
type Statement interface {
	stmtMarker()
}

// VariableDeclStmt represents `const x: T = expr` or `let x: T = expr`.
type VariableDeclStmt struct {
	Name           string
	Type           TypeNode // may be nil
	Mutable        bool     // const = false, let = true
	Init           Expression
	SourceLocation SourceLocation
}

func (VariableDeclStmt) stmtMarker() {}

// AssignmentStmt represents `target = value`.
type AssignmentStmt struct {
	Target         Expression
	Value          Expression
	SourceLocation SourceLocation
}

func (AssignmentStmt) stmtMarker() {}

// IfStmt represents an if/else statement.
type IfStmt struct {
	Condition      Expression
	Then           []Statement
	Else           []Statement // may be nil
	SourceLocation SourceLocation
}

func (IfStmt) stmtMarker() {}

// ForStmt represents a for loop with constant bounds.
type ForStmt struct {
	Init           VariableDeclStmt
	Condition      Expression
	Update         Statement
	Body           []Statement
	SourceLocation SourceLocation
}

func (ForStmt) stmtMarker() {}

// ReturnStmt represents a return statement.
type ReturnStmt struct {
	Value          Expression // may be nil
	SourceLocation SourceLocation
}

func (ReturnStmt) stmtMarker() {}

// ExpressionStmt represents an expression used as a statement.
type ExpressionStmt struct {
	Expr           Expression
	SourceLocation SourceLocation
}

func (ExpressionStmt) stmtMarker() {}

// ---------------------------------------------------------------------------
// Expressions
// ---------------------------------------------------------------------------

// Expression is the interface for all AST expression nodes.
type Expression interface {
	exprMarker()
}

// BinaryExpr represents a binary operation like `a + b`.
type BinaryExpr struct {
	Op    string // "+", "-", "*", "/", "%", "===", "!==", "<", "<=", ">", ">=", "&&", "||", "&", "|", "^"
	Left  Expression
	Right Expression
}

func (BinaryExpr) exprMarker() {}

// UnaryExpr represents a unary operation like `!a`, `-a`, `~a`.
type UnaryExpr struct {
	Op      string // "!", "-", "~"
	Operand Expression
}

func (UnaryExpr) exprMarker() {}

// CallExpr represents a function/method call.
type CallExpr struct {
	Callee Expression
	Args   []Expression
}

func (CallExpr) exprMarker() {}

// MemberExpr represents a member access like `obj.property` (not `this.x`).
type MemberExpr struct {
	Object   Expression
	Property string
}

func (MemberExpr) exprMarker() {}

// Identifier represents a variable or name reference.
type Identifier struct {
	Name string
}

func (Identifier) exprMarker() {}

// BigIntLiteral represents a bigint literal like `42n`.
type BigIntLiteral struct {
	Value int64
}

func (BigIntLiteral) exprMarker() {}

// BoolLiteral represents a boolean literal.
type BoolLiteral struct {
	Value bool
}

func (BoolLiteral) exprMarker() {}

// ByteStringLiteral represents a hex-encoded byte string literal.
type ByteStringLiteral struct {
	Value string // hex-encoded
}

func (ByteStringLiteral) exprMarker() {}

// TernaryExpr represents a conditional expression: `cond ? a : b`.
type TernaryExpr struct {
	Condition  Expression
	Consequent Expression
	Alternate  Expression
}

func (TernaryExpr) exprMarker() {}

// PropertyAccessExpr represents `this.x` (property access on the contract).
type PropertyAccessExpr struct {
	Property string
}

func (PropertyAccessExpr) exprMarker() {}

// IndexAccessExpr represents `arr[i]`.
type IndexAccessExpr struct {
	Object Expression
	Index  Expression
}

func (IndexAccessExpr) exprMarker() {}

// IncrementExpr represents `i++` or `++i`.
type IncrementExpr struct {
	Operand Expression
	Prefix  bool
}

func (IncrementExpr) exprMarker() {}

// DecrementExpr represents `i--` or `--i`.
type DecrementExpr struct {
	Operand Expression
	Prefix  bool
}

func (DecrementExpr) exprMarker() {}

// ---------------------------------------------------------------------------
// Primitive type names
// ---------------------------------------------------------------------------

var primitiveTypeNames = map[string]bool{
	"bigint":         true,
	"boolean":        true,
	"ByteString":     true,
	"PubKey":         true,
	"Sig":            true,
	"Sha256":         true,
	"Ripemd160":      true,
	"Addr":           true,
	"SigHashPreimage": true,
	"RabinSig":       true,
	"RabinPubKey":    true,
	"void":           true,
}

// IsPrimitiveType returns true if the name is a recognized TSOP primitive type.
func IsPrimitiveType(name string) bool {
	return primitiveTypeNames[name]
}

File: compilers/rust/src/frontend/ast.rs
//! Rúnar AST types.
//!
//! These types mirror the TypeScript `runar-ast.ts` definitions. They represent
//! the parsed contract structure before ANF lowering.

// ---------------------------------------------------------------------------
// Source locations
// ---------------------------------------------------------------------------

/// Source location in the original file.
#[derive(Debug, Clone)]
pub struct SourceLocation {
    pub file: String,
    pub line: usize,
    pub column: usize,
}

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/// Primitive type names recognized by Rúnar.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum PrimitiveTypeName {
    Bigint,
    Boolean,
    ByteString,
    PubKey,
    Sig,
    Sha256,
    Ripemd160,
    Addr,
    SigHashPreimage,
    RabinSig,
    RabinPubKey,
    Point,
    Void,
}

impl PrimitiveTypeName {
    /// Parse a string into a PrimitiveTypeName, if recognized.
    pub fn from_str(s: &str) -> Option<PrimitiveTypeName> {
        match s {
            "bigint" => Some(PrimitiveTypeName::Bigint),
            "boolean" => Some(PrimitiveTypeName::Boolean),
            "ByteString" => Some(PrimitiveTypeName::ByteString),
            "PubKey" => Some(PrimitiveTypeName::PubKey),
            "Sig" => Some(PrimitiveTypeName::Sig),
            "Sha256" => Some(PrimitiveTypeName::Sha256),
            "Ripemd160" => Some(PrimitiveTypeName::Ripemd160),
            "Addr" => Some(PrimitiveTypeName::Addr),
            "SigHashPreimage" => Some(PrimitiveTypeName::SigHashPreimage),
            "RabinSig" => Some(PrimitiveTypeName::RabinSig),
            "RabinPubKey" => Some(PrimitiveTypeName::RabinPubKey),
            "Point" => Some(PrimitiveTypeName::Point),
            "void" => Some(PrimitiveTypeName::Void),
            _ => None,
        }
    }

    /// Convert back to string representation.
    pub fn as_str(&self) -> &'static str {
        match self {
            PrimitiveTypeName::Bigint => "bigint",
            PrimitiveTypeName::Boolean => "boolean",
            PrimitiveTypeName::ByteString => "ByteString",
            PrimitiveTypeName::PubKey => "PubKey",
            PrimitiveTypeName::Sig => "Sig",
            PrimitiveTypeName::Sha256 => "Sha256",
            PrimitiveTypeName::Ripemd160 => "Ripemd160",
            PrimitiveTypeName::Addr => "Addr",
            PrimitiveTypeName::SigHashPreimage => "SigHashPreimage",
            PrimitiveTypeName::RabinSig => "RabinSig",
            PrimitiveTypeName::RabinPubKey => "RabinPubKey",
            PrimitiveTypeName::Point => "Point",
            PrimitiveTypeName::Void => "void",
        }
    }
}

/// A type node in the AST.
#[derive(Debug, Clone)]
pub enum TypeNode {
    Primitive(PrimitiveTypeName),
    FixedArray {
        element: Box<TypeNode>,
        length: usize,
    },
    Custom(String),
}

// ---------------------------------------------------------------------------
// Top-level nodes
// ---------------------------------------------------------------------------

/// A complete Rúnar contract.
#[derive(Debug, Clone)]
pub struct ContractNode {
    pub name: String,
    pub parent_class: String, // "SmartContract" or "StatefulSmartContract"
    pub properties: Vec<PropertyNode>,
    pub constructor: MethodNode,
    pub methods: Vec<MethodNode>,
    pub source_file: String,
}

/// A contract property declaration.
#[derive(Debug, Clone)]
pub struct PropertyNode {
    pub name: String,
    pub prop_type: TypeNode,
    pub readonly: bool,
    pub initializer: Option<Expression>,
    pub source_location: SourceLocation,
}

/// A method (constructor or named method).
#[derive(Debug, Clone)]
pub struct MethodNode {
    pub name: String,
    pub params: Vec<ParamNode>,
    pub body: Vec<Statement>,
    pub visibility: Visibility,
    pub source_location: SourceLocation,
}

/// Method visibility.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Visibility {
    Public,
    Private,
}

/// A method parameter.
#[derive(Debug, Clone)]
pub struct ParamNode {
    pub name: String,
    pub param_type: TypeNode,
}

// ---------------------------------------------------------------------------
// Statements
// ---------------------------------------------------------------------------

/// Statement variants.
#[derive(Debug, Clone)]
pub enum Statement {
    VariableDecl {
        name: String,
        var_type: Option<TypeNode>,
        mutable: bool,
        init: Expression,
        source_location: SourceLocation,
    },
    Assignment {
        target: Expression,
        value: Expression,
        source_location: SourceLocation,
    },
    IfStatement {
        condition: Expression,
        then_branch: Vec<Statement>,
        else_branch: Option<Vec<Statement>>,
        source_location: SourceLocation,
    },
    ForStatement {
        init: Box<Statement>, // Always VariableDecl
        condition: Expression,
        update: Box<Statement>,
        body: Vec<Statement>,
        source_location: SourceLocation,
    },
    ReturnStatement {
        value: Option<Expression>,
        source_location: SourceLocation,
    },
    ExpressionStatement {
        expression: Expression,
        source_location: SourceLocation,
    },
}

// ---------------------------------------------------------------------------
// Expressions
// ---------------------------------------------------------------------------

/// Binary operator.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum BinaryOp {
    Add,       // +
    Sub,       // -
    Mul,       // *
    Div,       // /
    Mod,       // %
    StrictEq,  // ===
    StrictNe,  // !==
    Lt,        // <
    Le,        // <=
    Gt,        // >
    Ge,        // >=
    And,       // &&
    Or,        // ||
    BitAnd,    // &
    BitOr,     // |
    BitXor,    // ^
    Shl,       // <<
    Shr,       // >>
}

impl BinaryOp {
    /// Convert to the string representation used in ANF IR.
    pub fn as_str(&self) -> &'static str {
        match self {
            BinaryOp::Add => "+",
            BinaryOp::Sub => "-",
            BinaryOp::Mul => "*",
            BinaryOp::Div => "/",
            BinaryOp::Mod => "%",
            BinaryOp::StrictEq => "===",
            BinaryOp::StrictNe => "!==",
            BinaryOp::Lt => "<",
            BinaryOp::Le => "<=",
            BinaryOp::Gt => ">",
            BinaryOp::Ge => ">=",
            BinaryOp::And => "&&",
            BinaryOp::Or => "||",
            BinaryOp::BitAnd => "&",
            BinaryOp::BitOr => "|",
            BinaryOp::BitXor => "^",
            BinaryOp::Shl => "<<",
            BinaryOp::Shr => ">>",
        }
    }
}

/// Unary operator.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum UnaryOp {
    Not,    // !
    Neg,    // -
    BitNot, // ~
}

impl UnaryOp {
    /// Convert to the string representation used in ANF IR.
    pub fn as_str(&self) -> &'static str {
        match self {
            UnaryOp::Not => "!",
            UnaryOp::Neg => "-",
            UnaryOp::BitNot => "~",
        }
    }
}

/// Expression variants.
#[derive(Debug, Clone)]
pub enum Expression {
    BinaryExpr {
        op: BinaryOp,
        left: Box<Expression>,
        right: Box<Expression>,
    },
    UnaryExpr {
        op: UnaryOp,
        operand: Box<Expression>,
    },
    CallExpr {
        callee: Box<Expression>,
        args: Vec<Expression>,
    },
    MemberExpr {
        object: Box<Expression>,
        property: String,
    },
    Identifier {
        name: String,
    },
    BigIntLiteral {
        value: i64,
    },
    BoolLiteral {
        value: bool,
    },
    ByteStringLiteral {
        value: String, // hex-encoded
    },
    TernaryExpr {
        condition: Box<Expression>,
        consequent: Box<Expression>,
        alternate: Box<Expression>,
    },
    PropertyAccess {
        property: String, // this.x -> property = "x"
    },
    IndexAccess {
        object: Box<Expression>,
        index: Box<Expression>,
    },
    IncrementExpr {
        operand: Box<Expression>,
        prefix: bool,
    },
    DecrementExpr {
        operand: Box<Expression>,
        prefix: bool,
    },
    ArrayLiteral {
        elements: Vec<Expression>,
    },
}
